divintseg/_core.py

```diff
@@ -422,3 +422,151 @@ def similarity(
+    ... )
+    ...
+    ... df
+         REGION    SUBREGION    S    T
+    0  Region 1  Subregion A  100    0
+    1  Region 1  Subregion B   50   50
+    2  Region 2  Subregion C    0  100
+    3  Region 2  Subregion D    0   50
+    4  Region 2  Subregion E   10   90
+
+    >>> from divintseg import isolation
+    ...
+    ... isolation(df, "S", by="REGION", over="SUBREGION")
+         REGION        S
+    0  Region 1  0.83333
+    1  Region 2      0.1
+
+    Let's look at what this example computed. First, we have
+    to see how likely each person in group S is to see other
+    members of their own group in their subregion. This is as
+    follows:
+
+    +----------+-------------+-----------------------+
+    | Region   | Subregion   | Likelihood of an S    |
+    +==========+=============+=======================+
+    | Region 1 | Subregion A | 100 / (100 + 0) = 1.0 |
+    +----------+-------------+-----------------------+
+    | Region 1 | Subregion B | 50 / (50 + 50) = 0.5  |
+    +----------+-------------+-----------------------+
+    | Region 2 | Subregion C | 0 / (0 + 100) = 0.0   |
+    +----------+-------------+-----------------------+
+    | Region 2 | Subregion D | 0 / (0 + 50) = 0.0    |
+    +----------+-------------+-----------------------+
+    | Region 2 | Subregion E | 10 / (10 + 90) = 0.1  |
+    +----------+-------------+-----------------------+
+
+    Next, we can compute the fraction of all S's in
+    each subregion of each region. There are 150 S's
+    in Region 1 and 10 S's in region 2, therefore, we have:
+
+    +----------+-------------+--------------------------------+
+    | Region   | Subregion   | Fraction of all As in Region   |
+    +==========+=============+================================+
+    | Region 1 | Subregion A | 100 / 150 = 0.6667             |
+    +----------+-------------+--------------------------------+
+    | Region 1 | Subregion B | 50 / 150 = 0.3333              |
+    +----------+-------------+--------------------------------+
+    | Region 2 | Subregion C | 0 / 10 = 0.0000                |
+    +----------+-------------+--------------------------------+
+    | Region 2 | Subregion D | 0 / 10 = 0.0000                |
+    +----------+-------------+--------------------------------+
+    | Region 2 | Subregion E | 10 / 10 = 1.0000               |
+    +----------+-------------+--------------------------------+
+
+    Finally, for each subregion, we multiply these together and
+    add them up the values for the subregions in each region.
+    For Region 1, we get
+
+    .. math::
+        (0.6667 * 1.0) + (0.3333 * 0.5) = 0.8333.
+
+    For Region 2, we get
+
+    .. math::
+        0.0 * 0.0 + 0.0 * 0.0 + 1.000 * 0.1 = 0.1.
+
+    Note that the implentation may not do this math exactly as
+    specified here, but it will do something equivalent.
+    """
+    if by is None:
+        df_communities["REGION"] = "Region " + df_communities.index.astype(str)
```

Comment:
Since this is a library function that people are potentially going to call from a lot of different places, I'd rather try to avoid modifying the data frame that was passed in. As a user of the library, I would generally be surprised if a data frame I passed in was modified.

That's general guidance for this kind of code, but in this case, let's just make `by` and `over` non-optional as mentioned in the comment above and avoid the problem that way.